File: tests/isapprox.jl
using DelimitedFiles, ArgParse, Measurements


function parse_commandline()
    s = ArgParseSettings()

    @add_arg_table! s begin
        "--col_A"
            help = "column in file A to compare"
            arg_type = Int
            default = 1
        "--col_err_A"
            help = "corresponding error in file A"
            arg_type = Int
        "--col_B"
            help = "column in file B to compare"
            arg_type = Int
            default = 1
        "--col_err_B"
            help = "corresponding error in file B"
            arg_type = Int
        "--verbose", "-v"
            help = "be verbose"
            action = :store_true
        "fileA"
            help = "File A"
            arg_type = String
            required = true
        "fileB"
            help = "File B"
            arg_type = String
            required = true
    end

    return parse_args(s)
end

function main()
    parse_args = parse_commandline()

    file_a = parse_args["fileA"]
    file_b = parse_args["fileB"]

    a_data = readdlm(file_a, comments=true)
    b_data = readdlm(file_b, comments=true)

    col_a = parse_args["col_A"]
    col_err_a = parse_args["col_err_A"]
    col_b = parse_args["col_B"]
    col_err_b = parse_args["col_err_B"]

    a = measurement.(a_data[:,col_a], (isnothing(col_err_a) ? 0 : a_data[:,col_err_a]))
    b = measurement.(b_data[:,col_b], (isnothing(col_err_b) ? 0 : b_data[:,col_err_b]))

    N = size(a)[1]

    if N != size(b)[1]
        @error "Column dimensions in file A and B differ"
    end

    println("Comparing column $col_a from $file_a and column $col_b from $file_b:")
    if a == b
        println("Colums are equal")
        return
    end

    zscore = stdscore.(a,b)

    for i = 1:N
        z = zscore[i]
        if (parse_args["verbose"] || abs(z) > 1)
            println("Row $i: a = $(a[i]), b = $(b[i]), zscore = $z")
        end
    end
    println("done.")
end

main()

File: tests/isapprox.py
import argparse
import numpy as np

# Helper function for measurements
def zscore(a_val, a_err, b_val, b_err):
    err = np.sqrt(a_err**2 + b_err**2)
    return (a_val - b_val)/(err)

def parse_commandline():
    parser = argparse.ArgumentParser(description="Compare columns in two files with optional error columns.")

    parser.add_argument("--col_A", type=int, default=1, help="Column in file A to compare (1-based index).")
    parser.add_argument("--col_err_A", type=int, help="Column for error in file A (1-based index).")
    parser.add_argument("--col_B", type=int, default=1, help="Column in file B to compare (1-based index).")
    parser.add_argument("--col_err_B", type=int, help="Column for error in file B (1-based index).")
    parser.add_argument("--verbose", "-v", action="store_true", help="Be verbose.")
    parser.add_argument("fileA", type=str, help="Path to file A.")
    parser.add_argument("fileB", type=str, help="Path to file B.")

    return parser.parse_args()

def main():
    args = parse_commandline()

    # Load data from files
    file_a = np.genfromtxt(args.fileA, comments="#")
    file_b = np.genfromtxt(args.fileB, comments="#")

    col_a = args.col_A - 1  # Convert to 0-based index
    col_err_a = args.col_err_A - 1 if args.col_err_A else None
    col_b = args.col_B - 1  # Convert to 0-based index
    col_err_b = args.col_err_B - 1 if args.col_err_B else None

    # Create measurements with or without errors
    a_errors = file_a[:, col_err_a] if col_err_a is not None else np.zeros(file_a.shape[0])
    b_errors = file_b[:, col_err_b] if col_err_b is not None else np.zeros(file_b.shape[0])

    a_vals = file_a[:, col_a]
    b_vals = file_b[:, col_b]

    if len(a_vals) != len(b_vals):
        print("Error: Column dimensions in file A and file B differ.")
        return

    print(f"Comparing column {col_a + 1} from {args.fileA} and column {col_b + 1} from {args.fileB}:")

    if np.array_equal(a_vals, b_vals):
        print("Columns are equal.")
        return

    # Compute z-scores
    zscores = zscore(a_vals, a_errors, b_vals, b_errors)

    for i, z in enumerate(zscores):
        if args.verbose or z > 1:
            print(f"Row {i + 1}: a = {a_vals[i]}, b = {b_vals[i]}, zscore = {z}")

    print("done.")

if __name__ == "__main__":
    main()
